### PR TITLE
Loadouts V3

### LINKED ISSUE #671: More Loadout Features and Changes needed.
The following changes would be positive for loadouts:

Make subcategories for similar loadout items (Cigarettes, Gloves, Survival gear, etc.) in the Loadout tabs. I'm thinking dropdown menus would work great here. [XAML + C#]

Make a server config option for Loadout Items "Exclusive" behavior, where if you select one item in a loadout it'll lockout or hide other similar items (you can only select one jumpsuit, pair of gloves, etc.) [YAML Datafield + Datafield Matching checks in C#]

Make Loadouts save on a character by job basis. Where you can make and save different job loadouts for different characters. DeltaV/Frontier Station loadout buttons and their character by job loadout system is a good example here.
[C# Serverside database saving for user characters' loadouts]

### LINKED ISSUE #691: "Job-specific items are tedious to select and unselect when switching roles"
# Description

Two parts:

1. it's not clear to new players how to regain their lost points when they switch jobs, having to go to "show unusable" and click it is unintuitive apparently

2. players who *do* know how to do it; say it's tedious to have to click several items and navigate the huge "show unusable" menu to be able to switch yourself to any other role.

### LINKED ISSUE #708: Request: Allowing for "OR" CharacterRequirements
i.e

To play Captain:

You have to be over 20; OR a Human OR have the "Masochist" trait.

### LINKED ISSUE #723: Loadouts / Traits / Character Requirements mega-issue
# Description

As requested and spoken about:

## Problems:

- [ ] test fail should be thrown if using an invalid ID inside a loadout entry
- [ ] Categories will overflow the UI if there's too many, or a user's screen is too small
- [ ] Selecting an item you fit the requirements for; then failing to meet the requirements should de-select the item (or something similar)
- [ ] Traits / Loadouts will reset scroll upon any update
- [ ] Traits / Loadouts you do not meet the requirements for; that you then switch to meet the requirements will not update the entry until Loadouts / traits is updated
i.e if I'm an Oni, and I want a Felinid trait so I switch to Felinid; the trait menu won't update until I do something to update it 

## Problems (needs replicating or testing)

- [x]  Whitelist doesn't work
- [ ]  JobBans don't work
- [ ]  AntagBans don't work (Wizden issue, there's a PR open to fix this but I doubt it'd work for us)
- [ ] Weight on the CharacterProfile and the Entity don't match up. (ask MnemoTechnician)

### Requested Features

- [ ] OrRequirements
i.e I need to be a Human, OR a Felinid to be Captain
- [ ] Requirements
	- [ ] Height
	- [ ] Width
	- [ ] Weight
	- [ ] Gender
	- [ ] Sex
	- [ ] Markings
- [ ] https://github.com/Simple-Station/Einstein-Engines/pull/643

I will add more as I remember them, it's quite late

### PR DESCRIPTION
**Pull request by DEATHB4DEFEAT**
# Description

Resolves #723
Resolves #708
Resolves #691
Resolves #671
Resolves #643

Adding/removing loadouts and categories (and trait stuff) will require disconnecting and reconnecting to the server (or anything that will reload the lobby UI) due to how I've done my caching.

---

# TODO

- [x] Test entity IDs from loadouts
- [x] Better Tab Container
	- [x] Scroll props
	- [x] Margin props
	- [x] Make vertical work
	- [x] Merge buttons
- [x] Subcategories
	- [x] Allow the same for traits (their tab is completely broken)
- [x] Give the Job Picker tab a BoxContainer
- [x] Different selected unusable color
- [x] Deselect unusable button
- Sort menu/Reverse sort direction button
- [x] Fix scroll resetting on updating
	- [x] Cache loadouts
- [x] Ensure everything that can affect reqs updates everything that checks reqs
- [x] More Requirements
	- [x] Height
	- [x] Width
	- [x] Weight
	- [x] Gender
	- [x] Sex
	- Markings
	- [x] Logic
		- [x] And
		- [x] Or
		- [x] Xor
	- [x] Item groups
- [x] #643
- Something better than `Exclusive`
	- Allow picking slots for things to go in (hands and inventories too)
- [x] Add a background to the loadout spriteView
- [x] Get better style colors
---
- [ ] More PRs
	- [ ] Remove duplicate command loadouts that are outside the command folders/categories
	- [ ] Categorize job loadouts (By dept?)
	- [ ] Fix [every loadout](https://discord.com/channels/1218698320155906090/1218698321053356060/1279260210208440361)'s IDs (who reviewed them??)
		- [ ] Make loadout ID migrations?

---

<details><summary><h1>Media</h1></summary>
<p>

## New loadouts/traits UI

image

image

### https://youtu.be/krg8TJoQG3I

image

image

</p>
</details>

---

# Changelog

:cl:
- fix: The loadout/trait editors' performance has been improved by 100~1000x
- tweak: The character editor's tabs look, feel, and function much better
- fix: Unusable loadouts/traits have an orange color when selected
- tweak: Loadout previews are now inside the button
- add: Added a button to remove any unusable loadouts (or traits) you have selected
- add: Loadouts and traits can have subcategories, though only Command loadouts use them at the moment
- fix: Fixed empty loadout categories not hiding
- fix: Fixed spacing on some loadout requirements' reason text
- fix: Fixed traits not updating after saving

### CODE REVIEW

**Comment by DEATHB4DEFEAT** [comment]
image

**Review by VMSolidus — COMMENTED** [review]
https://github.com/user-attachments/assets/a8e6385f-042f-4ff9-8452-5da1b6ee3891

This isn't a proper review yet, I am merely noting it here that I have verified through ingame testing(video provided) that this PR works as advertised. I will however need to go over this with a fine toothed comb, and also make my PR-PR that changes CowToolbox to serve as the "Extreme OR XOR shitton of new loadout requirements" testing tool, and also demonstrate to would-be-yml-contributors how the new system works.

File: Content.Client/UserInterface/Controls/NeoTabContainer.xaml.cs [hunk 1]
```diff
@@ -0,0 +1,299 @@
...
+
+    /// Sets the title of the tab associated with the given control
+    public void SetTabTitle(Control control, string title)
+    {
+        if (!_tabs.TryGetValue(control, out var button))
+            return;
+
+        if (button is Button b)
+            b.Text = title;
+    }
```
Comment (VMSolidus) [inline Content.Client/UserInterface/Controls/NeoTabContainer.xaml.cs:189]:
```suggestion
    {
        if (!_tabs.TryGetValue(control, out var button)
            || button is not Button b)
            return;

        b.Text = title;
    }
```

**Review by VMSolidus — APPROVED** [review]
Just this small change.

**Review by OldDanceJacket — APPROVED** [review]

**Pull request merged**